keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -202,6 +202,10 @@ For brevity - "old predicates" refers to now deprecated cloudprovider specific p
 * Scheduler will respect `Node.status.allocatable` and `Node.status.capacity` for CSI volumes if `CSINode` object is not available or has missing entry in `CSINode.spec.drivers[xyz].allocatable` during a deprecation period but kubelet will stop populating `Node.status.allocatable` and `Node.status.capacity` for CSI volumes.
   * After deprecation period for CSI volumes, limits coming from `Node.status.allocatable` and `Node.status.capacity` will be completely ignored by the scheduler.
   * After deprecation period, scheduler won't schedule any pods that use CSI volumes to a node with missing `CSINode` instance or missing driver in `CSINode.Spec.Drivers`.
+* To make sure the volume is from a CSI source, the scheduler may check whether a `CSIDriver` object with the provisioner's name exists.
```

Comment:
Can it take the StorageClass.Provisioner name and compare it to the CSINode.Driver.Name?

Reply:
It won't work to prevent scheduling of pods on nodes that don't have `CSINode` object yet.

Reply:
I am not sure if we want to support no scheduling if driver is not installed yet without a solution for autoscaling.  Otherwise this will be a regression in current behavior.

Reply:
> Can it take the StorageClass.Provisioner name and compare it to the CSINode.Driver.Name?

Yes, that's what I'm doing in my WIP PR #79429.

> I am not sure if we want to support no scheduling if driver is not installed yet without a solution for autoscaling. 

Should we drop that from the KEP until the autoscaler is updated?

Reply:
There is a section in KEP to cover that behaviour. May be we should update that section too?

Reply:
Removed. I kept it in the section for the deprecation period, line 204.

Reply:
> Can it take the StorageClass.Provisioner name and compare it to the CSINode.Driver.Name?

We couldn't do that because the volume might be from a real CSI driver that's not be installed in the node already.

This is a bigger problem for the "stop scheduling pods" case, so I think we have 2 options:

**a)** Change the autoscaler, update the scheduler to look at the `csidrivers` resource to count unbound CSI volumes and implement the "stop scheduling pods where driver is not installed" (this is mostly done in https://github.com/kubernetes/kubernetes/pull/79429) .

**b)** Simply compare the `storageclass.Provisioner` with `CSINode.Drivers[x].Name` to count unbound CSI volumes and **don't** implement the "stop scheduling pods" feature.

It sounds like the best option for now is **#b**, specially because we don't know when we'll be able to work on the autoscaler. Hopefully the autoscaler will be ready by the time the deprecation period ends, then we can look into

Does that sound good to you?

Reply:
I'm still not sure why we need the csidriver object? It should be sufficient to just compare provisioner name with driver name from csinode.

Can we move all references to not scheduling a pod if the driver is not installed to a single section and say that we can't implement it until we have a solution to integrate with autoscaler

Reply:
Yeah, the `CSIDriver` part is only necessary for the "stop scheduling pods", so we don't need it for now.

> Can we move all references to not scheduling a pod if the driver is not installed to a single section and say that we can't implement it until we have a solution to integrate with autoscaler

Sure, sounds good. I'll update this PR tomorrow.

Reply:
Removed reference to CSIDriver and clarified "not scheduling pods if driver is not installed" part.

Reply:
Do we still need to modify anything to talk about how topology is handled (for unbound PVCs with delayed binding)?

Reply:
As long as we don't prevent sched pods, unbound PVCs with delayed binding are handled just like the other cases. IMO we don't need to add that to the KEP, but let me know if you disagree.

For the context, this is how bound and unbound PVCs are handled without the "prevent sched pods". The predicate will:

1) Read the volume limits from CSINode
2) Count volumes by their limits key 
3) Check if there is a limit specified in the CSINode for each of these keys

In other words, we only block what's specified in the CSINode (blacklist), as opposed to *also* blocking everything that's *not* installed on the node (whitelist).